src/dataflow/core/type_inference.py

```diff
@@ -0,0 +1,330 @@
+            [cast(Type, AnonTypeVariable()) for arg in defn.args] if defn else []
+        )
+        ascribed_type = TypeApplication(
+            "Lambda", ascribed_arg_types + [ascribed_return_type]
+        )
+        comp_type = _unify(ascribed_type, defn_type, substitutions)
+        assert isinstance(
+            comp_type, TypeApplication
+        ), "unification of Lambdas should always produce a TypeApplication"
+        return Computation(
+            op, expression.id, rec_args, declared_type_args_list, comp_type
+        )
+
+    (roots, _) = roots_and_reentrancies(program)
+    assert (
+        len(roots) == 1
+    ), f"Expected Program to have a single root, got {roots} in {program}"
+    root_id = list(roots)[0]
+    return rec(id_to_expr[root_id])
+
+
+def _definition_to_type(
+    definition: Definition, declared_type_args: Dict[str, NamedTypeVariable]
+) -> Type:
+    return_type = _type_name_to_type(definition.type, declared_type_args)
+    arg_types = [_type_name_to_type(arg, declared_type_args) for arg in definition.args]
+
+    return TypeApplication("Lambda", arg_types + [return_type])
+
+
+def _type_name_to_type(
+    name: TypeName, declared_type_args: Dict[str, NamedTypeVariable]
+) -> Type:
+    if name is None:
+        return AnonTypeVariable()
+    if len(name.type_args) == 0:
+        if name.base in declared_type_args:
+            return declared_type_args[name.base]
+        else:
+            return TypeApplication(name.base, [])
+    else:
+        args = [_type_name_to_type(arg, declared_type_args) for arg in name.type_args]
+        return TypeApplication(name.base, args)
+
+
+def _type_to_type_name(t: Type) -> TypeName:
+    assert isinstance(
+        t, TypeApplication
+    ), f"_type_to_type_name expects a Type that is fully instantiated with no variables, but got {t}"
+    return TypeName(t.constructor, [_type_to_type_name(a) for a in t.args])
+
+
+def _apply_substitutions(t: Type, substitutions: Dict[TypeVariable, Type]):
+    if isinstance(t, TypeVariable):
+        if t in substitutions:
+            return _apply_substitutions(substitutions[t], substitutions)
+        else:
+            return t
+    elif isinstance(t, TypeApplication):
+        return TypeApplication(
+            t.constructor, [_apply_substitutions(arg, substitutions) for arg in t.args]
+        )
+
+    else:
+        raise TypeInferenceError(f"Unknown type {t}")
+
+
+def _unify(t1: Type, t2: Type, substitutions: Dict[TypeVariable, Type]) -> Type:
+    """See e.g. section 2.2 of
+    https://course.ccs.neu.edu/cs4410sp19/lec_type-inference_notes.html"""
+    t1 = _apply_substitutions(t1, substitutions)
+    t2 = _apply_substitutions(t2, substitutions)
+    if t1 == t2:
+        return t1
+    # If either t1 or t2 is a type variable, and it does not occur in the other type,
+    # then produce the substitution that binds it to the other type.
+    elif isinstance(t2, TypeVariable) and not isinstance(t1, TypeVariable):
```

Comment:
In Python, with `return`, it is more common to directly use `if` without `elif` or `else`.

```suggestion
    if isinstance(t2, TypeVariable) and not isinstance(t1, TypeVariable):
```